### PR TITLE
[WIP] Adding CPS project locking around restore commits

### PR DESCRIPTION
**Pull request by emgarten**
Adding CPS project locking around restore commits

Fixes https://github.com/NuGet/Home/issues/3807

//cc @drewgil @rrelyea

### CODE REVIEW

File: src/NuGet.Clients/PackageManagement.VisualStudio/ProjectSystems/CpsPackageReferenceProject.cs [hunk 1]
```diff
@@ -249,7 +250,21 @@ private static PackageReference ToPackageReference(LibraryDependency library, Nu
...
+        {
+            return await ThreadHelper.JoinableTaskFactory.RunAsync(async () =>
+            {
+                await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
+
+                var hierarchy = VsHierarchyUtility.ToVsHierarchy(EnvDTEProject);
+
+                var projectLock = ProjectHelper.GetProjectLock(_envDTEProject, hierarchy, _projectFullPath);
+
+                return Task.FromResult<INuGetLock>(projectLock);
```
Comment (alpaix) [inline src/NuGet.Clients/PackageManagement.VisualStudio/ProjectSystems/CpsPackageReferenceProject.cs]: simply 

``` c#
 return projectLock;
```

File: src/NuGet.Clients/VisualStudio.Facade/ProjectSystem/CPSProjectLock.cs [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+using System;
+using EnvDTE;
+using Microsoft.VisualStudio.Shell.Interop;
+using Microsoft.VisualStudio;
+using Microsoft.VisualStudio.ProjectSystem;
+using Microsoft.VisualStudio.Shell;
+using MsBuildProject = Microsoft.Build.Evaluation.Project;
+using Task = System.Threading.Tasks.Task;
+using NuGet.Common;
+using System.Threading;
```
Comment (alpaix) [inline src/NuGet.Clients/VisualStudio.Facade/ProjectSystem/CPSProjectLock.cs:13]: sort

File: src/NuGet.Clients/VisualStudio.Facade/ProjectSystem/CPSProjectLock.cs [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+#if VS14
+using Microsoft.VisualStudio.ProjectSystem.Designers;
+#elif VS15
+using Microsoft.VisualStudio.ProjectSystem.Properties;
+#endif
+
+namespace NuGet.VisualStudio.Facade.ProjectSystem
+{
+    internal class CPSProjectLock : INuGetLock
```
Comment (alpaix) [inline src/NuGet.Clients/VisualStudio.Facade/ProjectSystem/CPSProjectLock.cs:23]: is it actually `LegacyCPSProjectLock`?

File: src/NuGet.Clients/VisualStudio.Facade/NuGet.VisualStudio.Facade.csproj [hunk 4]
```diff
@@ -23,6 +23,7 @@
     <Reference Include="System.ComponentModel.Composition" />
   </ItemGroup>
   <ItemGroup>
+    <Compile Include="ProjectSystem\CPSProjectLock.cs" />
```
Comment (alpaix) [inline src/NuGet.Clients/VisualStudio.Facade/NuGet.VisualStudio.Facade.csproj:26]: Why can't it be conditionally included in `PackageManagement.VisualStudio`?

File: src/NuGet.Clients/VisualStudio.Facade/ProjectSystem/CPSProjectLock.cs [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+            });
+        }
+
+        public async Task ReleaseAsync()
+        {
+            await ThreadHelper.JoinableTaskFactory.RunAsync(async () =>
+            {
+                await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
+
+                // Since all calls happen on the UI thread no additional locking is needed
```
Comment (alpaix) [inline src/NuGet.Clients/VisualStudio.Facade/ProjectSystem/CPSProjectLock.cs]: Can you add a short doc comment on top of the class saying it's thread-safe by design?

File: src/NuGet.Clients/VisualStudio.Facade/ProjectSystem/CPSProjectLock.cs [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+{
+    internal class CPSProjectLock : INuGetLock
+    {
+        private readonly Project _project;
+        private readonly IVsHierarchy _hierarchy;
+        private ProjectWriteLockReleaser _lockReleaser;
+        private bool _isLocked;
+        private readonly string _projectPath;
+
+        public CPSProjectLock(Project project, IVsHierarchy hierarchy, string projectPath)
```
Comment (alpaix) [inline src/NuGet.Clients/VisualStudio.Facade/ProjectSystem/CPSProjectLock.cs:29]: null check for all 3 arguments

File: src/NuGet.Clients/VisualStudio.Facade/ProjectSystem/CPSProjectLock.cs [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+        {
+            await ThreadHelper.JoinableTaskFactory.RunAsync(async () =>
+            {
+                await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
+
+                // Noop if this was already called
+                if (!_isLocked)
+                {
+                    var vsProject = (IVsProject)_hierarchy;
+                    UnconfiguredProject unconfiguredProject = ProjectHelper.GetUnconfiguredProject(vsProject);
```
Comment (alpaix) [inline src/NuGet.Clients/VisualStudio.Facade/ProjectSystem/CPSProjectLock.cs]: use `var`

File: src/NuGet.Clients/VisualStudio.Facade/ProjectSystem/CPSProjectLock.cs [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+                    if (unconfiguredProject != null)
```
Comment (alpaix) [inline src/NuGet.Clients/VisualStudio.Facade/ProjectSystem/CPSProjectLock.cs]: I'd suggest throwing `InvalidOperationException` here

File: src/NuGet.Clients/VisualStudio.Facade/ProjectSystem/CPSProjectLock.cs (hunk 3 above)
Comment (alpaix) [reply src/NuGet.Clients/VisualStudio.Facade/ProjectSystem/CPSProjectLock.cs:23]: To clarify the purpose of it `LegacyCPSProjectWriteLock`

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreRunner.cs [hunk 9]
```diff
@@ -19,22 +19,41 @@ public static class RestoreRunner
         /// <summary>
         /// Create requests, execute requests, and commit restore results.
         /// </summary>
-        public static async Task<IReadOnlyList<RestoreSummary>> Run(RestoreArgs restoreContext)
+        public static Task<IReadOnlyList<RestoreSummary>> Run(RestoreArgs restoreContext)
```
Comment (alpaix) [inline src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreRunner.cs:22]: `RunAsync`?

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreRunner.cs [hunk 10]
```diff
@@ -19,22 +19,41 @@ public static class RestoreRunner
...
             IEnumerable<RestoreSummaryRequest> restoreRequests,
             RestoreArgs restoreContext)
         {
+            return Run(restoreRequests, restoreContext, Enumerable.Empty<INuGetLock>());
+        }
+
+        /// <summary>
+        /// Execute and commit restore requests.
+        /// </summary>
+        public static async Task<IReadOnlyList<RestoreSummary>> Run(
```
Comment (alpaix) [inline src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreRunner.cs:52]: `RunAsync`

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreRunner.cs [hunk 11]
```diff
@@ -216,11 +237,41 @@ private static int GetMaxTaskCount(RestoreArgs restoreContext)
             return maxTasks;
         }
 
-        private static async Task<RestoreSummary> ExecuteAndCommit(RestoreSummaryRequest summaryRequest)
+        private static async Task<RestoreSummary> ExecuteAndCommit(
```
Comment (alpaix) [inline src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreRunner.cs:240]: `ExecuteAndCommitAsync`

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreRunner.cs [hunk 12]
```diff
@@ -216,11 +237,41 @@ private static int GetMaxTaskCount(RestoreArgs restoreContext)
...
         {
             var result = await Execute(summaryRequest);
 
-            return await Commit(result);
+            var locksForProject = GetLocksForProject(locks, summaryRequest.Request.Project);
+
+            return await CommitInLock(result, locksForProject);
+        }
+
+        public static async Task<RestoreSummary> CommitInLock(RestoreResultPair result, IEnumerable<INuGetLock> locks)
```
Comment (alpaix) [inline src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreRunner.cs]: `CommitInLockAsync`

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreRunner.cs [hunk 13]
```diff
@@ -216,11 +237,41 @@ private static int GetMaxTaskCount(RestoreArgs restoreContext)
...
+        {
+            var activeLocks = new List<INuGetLock>();
+
+            try
+            {
+                // Get locks
+                foreach (var current in locks)
```
Comment (alpaix) [inline src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreRunner.cs]: I'd suggest adding and using `CompositeLock` class to wrap it

File: src/NuGet.Core/NuGet.Common/INuGetLock.cs [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace NuGet.Common
+{
+    /// <summary>
+    /// Basic get/release interface for an Async lock.
+    /// </summary>
+    public interface INuGetLock : IDisposable
```
Comment (alpaix) [inline src/NuGet.Core/NuGet.Common/INuGetLock.cs]: 1. `IDisposable` is implementation specific and shouldn't be part of the interface declaration.
2. I can't find it's actually used anywhere.

File: src/NuGet.Core/NuGet.PackageManagement/BuildIntegration/DependencyGraphRestoreUtility.cs [hunk 15]
```diff
@@ -365,5 +371,17 @@ public static class DependencyGraphRestoreUtility
 
             return restoreContext;
         }
+
+        private static async Task<List<INuGetLock>> GetLocksFromSolution(ISolutionManager solutionManager)
```
Comment (alpaix) [inline src/NuGet.Core/NuGet.PackageManagement/BuildIntegration/DependencyGraphRestoreUtility.cs:372]: I'd suggest moving this method to `ISolutionManager`

File: src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs [hunk 16]
```diff
@@ -2575,7 +2575,17 @@ var projectUniqueNamesForBuildIntToUpdate
                 else
                 {
                     // Write out the lock file
-                    await RestoreRunner.Commit(projectAction.RestoreResultPair);
+                    var buildIntProject = (BuildIntegratedNuGetProject)projectAction.Project;
+                    var projectLock = await buildIntProject.GetProjectLockAsync();
+
```
Comment (alpaix) [inline src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs:2580]: Missing `AcquireAsync`?

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreRunner.cs [hunk 17]
```diff
@@ -357,5 +408,30 @@ public static string GetInvalidInputErrorMessage(string input)
 
             return Strings.Error_InvalidCommandLineInput;
         }
+
+        private static Dictionary<string, List<INuGetLock>> GetIdToLocks(IEnumerable<INuGetLock> locks)
```
Comment (alpaix) [inline src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreRunner.cs:394]: Multiple locks per project?

**Review by alpaix — CHANGES_REQUESTED** [review]

**Comment by alpaix** [comment]
I'd suggest improving the design by having a class like this

``` c#
public class ProjectSystemCompositeLock
{
    private readonly Dictionary<string, INuGetLock> locks;

    public INuGetLock GetProjectLock(string projectPath);
}
```

And also a method `ISolutionManager.GetProjectSystemLock` returning an instance.

Thoughts?

**Comment by emgarten** [comment]
`ProjectSystemCompositeLock` could be nice, it would make it more obvious how to find the lock for a specific project. Also, yes I think allowing multiple per project makes sense since here the lock is for CPS, not for NuGet. If nuget wanted to add it's own machine wide lock for each project to keep VS and nuget.exe from conflicting it could then be done by just adding another lock here.

I'd like to see the CPS lock actually solve the restore problem before doing any refactoring on this, this was just to make it possible.

File: src/NuGet.Clients/VisualStudio.Facade/ProjectSystem/CPSProjectLock.cs [hunk 18]
```diff
@@ -0,0 +1,72 @@
...
+                        using (ProjectWriteLockReleaser x = await service.WriteLockAsync())
+                        {
+                            // Run action within the lock
+                            return await asyncAction();
+                        }
+                    }
+                }
+
+                // Run without the lock if the service cannot be acquired
+                return await asyncAction();
```
Comment (joelverhagen) [inline src/NuGet.Clients/VisualStudio.Facade/ProjectSystem/CPSProjectLock.cs:68]: When could we not get the lock? This seems kind of risky to not understand.

Comment (emgarten) [reply src/NuGet.Clients/VisualStudio.Facade/ProjectSystem/CPSProjectLock.cs:68]: There may be scenarios where the project isn't loaded, in those cases the lock isn't needed.

**Pull request closed without merging**